### PR TITLE
Bug 1927017: Use file observer for Proxy CA changes

### PR DESCRIPTION
**Pull request by suhanime**
Config map controller observed changes for Proxy CA and restarted the
pod when it did. Whenever it saw a change in Proxy CA, it restarted the
pod without releasing the lease, so the new pod took about 8 minutes to
acquire the new lease.

In order to support a graceful shutdown, this commit also aims at
dropping the config map controller and instead observing the changes via
a file observer. Whenever there is a change in Proxy CA, the observer
will call cancel context of leader election kick-starting a graceful
shutdown, and restart the pod.

/assign @joelddiaz 
/cc @abhinavdahiya

### CODE REVIEW

File: pkg/cmd/operator/cmd.go [hunk 1]
```diff
@@ -113,6 +120,11 @@ func NewOperator() *cobra.Command {
 			ctx, cancel := context.WithCancel(context.Background())
 			defer cancel()
 
+			caConfigMapPath := filepath.Join(caConfigMapMountPath, caConfigMapName)
+			if _, err := os.Stat(caConfigMapPath); err != nil {
+				go terminateWhenProxyChanges(caConfigMapPath, cancel)
+			}
```
Comment (suhanime) [inline pkg/cmd/operator/cmd.go:126]: I think the place where I am calling the goroutine is not correct - any pointers?

File: pkg/cmd/operator/cmd.go [hunk 2]
```diff
@@ -113,6 +120,11 @@ func NewOperator() *cobra.Command {
 			ctx, cancel := context.WithCancel(context.Background())
 			defer cancel()
 
+			caConfigMapPath := filepath.Join(caConfigMapMountPath, caConfigMapName)
+			if _, err := os.Stat(caConfigMapPath); err == nil {
+				go terminateWhenProxyChanges(caConfigMapPath, cancel)
```
Comment (abhinavdahiya) [inline pkg/cmd/operator/cmd.go:125]: we already start the fileobserver in a go routine https://github.com/openshift/cloud-credential-operator/pull/320/files#diff-2f87b1f2687fa01b2bd650174a4d17234de24ce7061f469e90c3ea56d72a77ddR232-R236, so why start this function also in go routine?

File: pkg/cmd/operator/cmd.go [hunk 3]
```diff
@@ -113,6 +120,11 @@ func NewOperator() *cobra.Command {
 			ctx, cancel := context.WithCancel(context.Background())
 			defer cancel()
```
Comment (abhinavdahiya) [inline pkg/cmd/operator/cmd.go:122]: i would personally would have it seen like,

// create the file observer

// add reactor for proxy files

// run the file observer

File: pkg/cmd/operator/cmd.go [hunk 4]
```diff
@@ -113,6 +120,11 @@ func NewOperator() *cobra.Command {
...
 
+			caConfigMapPath := filepath.Join(caConfigMapMountPath, caConfigMapName)
+			if _, err := os.Stat(caConfigMapPath); err == nil {
```
Comment (abhinavdahiya) [inline pkg/cmd/operator/cmd.go:124]: also i would like us to comment why we are checking for file to exist to start the reactor?

File: pkg/cmd/operator/cmd.go (hunk 2 above)
Comment (suhanime) [reply pkg/cmd/operator/cmd.go]: there wasn't any specific need - changed

File: pkg/cmd/operator/cmd.go [hunk 5]
```diff
@@ -196,3 +208,30 @@ func (writer glogWriter) Write(data []byte) (n int, err error) {
...
+			log.Info("Proxy CA configmap change detected, restarting pod")
+			cancel()
+			return nil
+		},
+		fileContents,
+	)
+	stop := make(chan struct{})
+	go func() {
+		log.WithField("file", path).Info("running file observer")
+		obs.Run(stop)
```
[the remark this answers is not in the document]
Comment (suhanime) [reply pkg/cmd/operator/cmd.go]: nice catch - changed

**Review by joelddiaz — COMMENTED** [review]
/lgtm

**Pull request merged**